lib/agenda/index.ts

```diff
@@ -0,0 +1,156 @@
+import humanInterval from 'human-interval';
+import { EventEmitter } from 'events';
+import { MongoClient, Db as MongoDb } from 'mongodb';
+import { JobProcessingQueue } from './job-processing-queue';
+import { cancel } from './cancel';
+import { create } from './create';
+import { database } from './database';
+import { dbInit } from './db-init';
+import { defaultConcurrency } from './default-concurrency';
+import { defaultLockLifetime } from './default-lock-lifetime';
+import { defaultLockLimit } from './default-lock-limit';
+import { define } from './define';
+import { every } from './every';
+import { jobs } from './jobs';
+import { lockLimit } from './lock-limit';
+import { maxConcurrency } from './max-concurrency';
+import { mongo } from './mongo';
+import { name } from './name';
+import { now } from './now';
+import { processEvery } from './process-every';
+import { purge } from './purge';
+import { saveJob } from './save-job';
+import { schedule } from './schedule';
+import { sort } from './sort';
+import { start } from './start';
+import { stop } from './stop';
+import { findAndLockNextJob } from './find-and-lock-next-job';
+import { Job } from '../job';
+
+/**
+ * @class Agenda
+ * @param {Object} config - Agenda Config
+ * @param {Function} cb - Callback after Agenda has started and connected to mongo
+ * @property {Object} _name - Name of the current Agenda queue
+ * @property {Number} _processEvery
+ * @property {Number} _defaultConcurrency
+ * @property {Number} _maxConcurrency
+ * @property {Number} _defaultLockLimit
+ * @property {Number} _lockLimit
+ * @property {Object} _definitions
+ * @property {Object} _runningJobs
+ * @property {Object} _lockedJobs
+ * @property {Object} _jobQueue
+ * @property {Number} _defaultLockLifetime
+ * @property {Object} _sort
+ * @property {Object} _indices
+ * @property {Boolean} _isLockingOnTheFly
+ * @property {Array} _jobsToLock
+ */
+class Agenda extends EventEmitter {
+  _defaultConcurrency: any;
+  _defaultLockLifetime: any;
+  _defaultLockLimit: any;
+  _definitions: any;
+  _findAndLockNextJob: any;
+  _indices: any;
+  _isLockingOnTheFly: boolean;
+  _jobQueue: JobProcessingQueue;
+  _jobsToLock: Job[];
+  _lockedJobs: Job[];
+  _runningJobs: Job[];
+  _lockLimit: any;
+  _maxConcurrency: any;
+  _mongoUseUnifiedTopology?: boolean;
+  _name: any;
+  _processEvery: number;
+  _ready: Promise<unknown>;
+  _sort: any;
+  _db!: MongoClient;
+  _mdb!: MongoDb;
+  _collection: any;
+  _nextScanAt: any;
+  _processInterval: any;
```

Comment:
Can some of these `any`'s be `number`?

Reply:
Actually no, I think collection is a mongodb connection, nextScanAt is null or a Date, and processInterval is string or a number as processInterval can be afaik a number, a chron string or human readable string like "5 seconds"

Reply:
_processInterval is actually a timer .. therefore `_processInterval?: ReturnType<typeof setInterval>` would be correct.  _processEvery is the thing you were referring too, it's a string or a number.

feel free to look up types in my fork, I spended already some time to figure them out :-)
https://github.com/hokify/agenda/tree/typescript/src/types
https://github.com/hokify/agenda/blob/typescript/src/index.ts

Reply:
Wrapping a type in a generic like ReturnType makes it harder to read the actual type in an IDE I’d suggest against them.

Reply:
It's an internal type, everything is better than any ;)

Reply:
Using internal generics doesn't make it easier. It means that new dev not only now need to know typescript but also generics, etc. Just use the type. It's a number that `setInterval` returns, yes this varies in browsers, etc. we're writing a node lib so that doesn't matter.

Reply:
<img width="473" alt="Screen Shot 2020-10-15 at 7 46 05 pm" src="https://user-images.githubusercontent.com/6525926/96103321-17a4a100-0f1f-11eb-84fb-0aa48fdd4573.png">


For inexperienced devs this is going to be way more confusing seeing it show a `NodeJS.Timeout` on a setInterval especially when `ReturnType<typeof setTimeout>` also returns that type. See why generics aren't always helpful, yes technically it's returning the "correct" type but that doesn't mean it's the best one for devs to see.